File: backend/app/Models/RoomParticipant.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Carbon\Carbon;

class RoomParticipant extends Model
{
    use HasFactory, HasUuids;

    protected $fillable = [
        'room_id',
        'user_id',
        'joined_at',
    ];

    protected $casts = [
        'joined_at' => 'datetime',
    ];

    public $timestamps = false;

    /**
     * Validation rules for room participant
     */
    public static function validationRules(): array
    {
        return [
            'room_id' => 'required|uuid|exists:rooms,id',
            'user_id' => 'required|uuid|exists:users,id',
        ];
    }

    /**
     * Room this participation belongs to
     */
    public function room(): BelongsTo
    {
        return $this->belongsTo(Room::class);
    }

    /**
     * User participating in the room
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get how long the user has been in the room
     */
    public function getDurationInRoom(): string
    {
        $duration = Carbon::now()->diff($this->joined_at);
        
        if ($duration->days > 0) {
            return $duration->days . ' day' . ($duration->days > 1 ? 's' : '');
        } elseif ($duration->h > 0) {
            return $duration->h . ' hour' . ($duration->h > 1 ? 's' : '');
        } elseif ($duration->i > 0) {
            return $duration->i . ' minute' . ($duration->i > 1 ? 's' : '');
        } else {
            return 'Just joined';
        }
    }

    /**
     * Check if user is the room administrator
     */
    public function isAdministrator(): bool
    {
        return $this->room->administrator_id === $this->user_id;
    }

    /**
     * Boot method to handle model events
     */
    protected static function boot()
    {
        parent::boot();

        // Set joined_at timestamp when creating
        static::creating(function ($participant) {
            if (!$participant->joined_at) {
                $participant->joined_at = Carbon::now();
            }
        });
    }
}

File: ruby-backend/app/models/room_participant.rb
# RoomParticipant model - Compatible with Laravel room_participants table
require 'sequel'
require 'securerandom'

class RoomParticipant < Sequel::Model
  # Use existing Laravel table structure
  set_dataset :room_participants
  
  # Plugins for enhanced functionality
  plugin :validation_helpers
  
  # Allow setting id for UUID compatibility
  unrestrict_primary_key
  
  # UUID generation for primary key (Laravel compatibility)
  def before_create
    super
    self.id ||= SecureRandom.uuid
  end
  
  # Associations
  many_to_one :room, class: :Room, key: :room_id
  many_to_one :user, class: :User, key: :user_id
  
  # Validation rules (compatible with Laravel validation)
  def validate
    super
    validates_presence [:room_id, :user_id]
    # Note: room_id and user_id are UUIDs (strings), not integers in Laravel compatibility mode
    validates_unique [:room_id, :user_id], message: 'user is already a participant in this room'
  end
  
  # Class methods for validation rules (Laravel compatibility)
  def self.validation_rules
    {
      room_id: 'required|uuid|exists:rooms,id',
      user_id: 'required|uuid|exists:users,id'
    }
  end
  
  # Duration in room calculation (Laravel compatibility)
  def duration_in_room
    return 'Just joined' unless joined_at
    
    duration = Time.now - joined_at
    
    case duration
    when 0..59
      'Just joined'
    when 60..3599
      minutes = (duration / 60).to_i
      "#{minutes} minute#{'s' if minutes != 1}"
    when 3600..86399
      hours = (duration / 3600).to_i
      "#{hours} hour#{'s' if hours != 1}"
    else
      days = (duration / 86400).to_i
      "#{days} day#{'s' if days != 1}"
    end
  end
  
  # Check if user is room administrator (Laravel compatibility)
  def administrator?
    room.administrator_id == user_id
  end
  
  # Serialization for API responses (Laravel format compatibility)
  def to_hash
    {
      id: id,
      room_id: room_id,
      user_id: user_id,
      user: user&.to_hash,
      joined_at: joined_at&.iso8601,
      duration_in_room: duration_in_room,
      is_administrator: administrator?
    }
  end
  
  def to_json(*args)
    to_hash.to_json(*args)
  end
  
  # Hooks for model events (Laravel compatibility)
  def before_create
    self.joined_at ||= Time.now
    super
  end
end